arch/sim/src/sim/up_macho_init.c

```diff
@@ -0,0 +1,100 @@
+/****************************************************************************
+ * arch/sim/src/sim/up_macho_init.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <assert.h>
+#include <stdlib.h>
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+typedef void (*init_func_t)(int argc, const char *argv[],
+                            const char *envp[], const char *apple[]);
+extern init_func_t mod_init_func_start \
```

Comment:
if we have to resolve macOS C++ initialization in this way. It's better to let Ubuntu use the same approach too. So we don't need maintain two different solution for the same issue.

Reply:
it sounds like a reasonable thing to do.